contrib/backup/multicloud/signer/signature.go

```diff
@@ -0,0 +1,319 @@
+		if _, ok := sign.SignedHeaderValues[lowerCaseKey]; ok {
+			sign.SignedHeaderValues[lowerCaseKey] = append(sign.SignedHeaderValues[lowerCaseKey], v...)
+			continue
+		}
+
+		headers = append(headers, lowerCaseKey)
+		sign.SignedHeaderValues[lowerCaseKey] = v
+	}
+	sort.Strings(headers)
+
+	sign.signedHeaders = strings.Join(headers, ";")
+
+	headerValues := make([]string, len(headers))
+	for i, k := range headers {
+		headerValues[i] = k + ":" +
+			strings.Join(sign.SignedHeaderValues[k], ",")
+	}
+	stripExcessSpaces(headerValues)
+	sign.canonicalHeaders = strings.Join(headerValues, "\n")
+}
+
+// ************* TASK 1: CREATE A CANONICAL REQUEST *************
+func (sign *Signature) buildCanonicalString() {
+
+	sign.Request.URL.RawQuery = strings.Replace(sign.Query.Encode(), "+", "%20", -1)
+
+	canonicalURI := getURIPath(sign.Request.URL)
+
+	sign.canonicalString = strings.Join([]string{
+		// Step 1 is to define the verb (GET, POST, etc.) Already defined in the Request.
+		sign.Request.Method,
+		// Step 2: Create canonical URI--the part of the URI from domain to query
+		canonicalURI,
+		// Step 3: Create the canonical query string.Query string values must be URL-encoded
+		// (space=%20). The parameters must be sorted by name.
+		sign.Request.URL.RawQuery,
+		// Step 4: Create the canonical headers and signed headers. Header names must be
+		// trimmed and lowercase, and in alpha order \n.
+		sign.canonicalHeaders + "\n",
+		// Step 5: Create the list of signed headers. This lists the headers in the
+		// canonical_headers list, delimited with ";" and in alpha order.
+		sign.signedHeaders,
+		// Step 6: Create payload hash (hash of the request body content). For GET requests,
+		// the payload is an empty string ("").
+		sign.requestPayload,
+	}, "\n") //Step 7: Combine elements to create canonical request
+	log.V(5).Infof("sign.Request.URL.RawQuery:%v", sign.Request.URL.RawQuery)
+	log.V(5).Infof("sign.canonicalHeaders:%v", sign.canonicalHeaders)
+	log.V(5).Infof("sign.signedHeaders:%v", sign.signedHeaders)
+	log.V(5).Infof("sign.requestPayload:%v", sign.requestPayload)
+	log.V(5).Infof("sign.canonicalString:%v", sign.canonicalString)
+}
+
+// ************* TASK 2: CREATE THE STRING TO SIGN *************
+// Match the algorithm to the hashing algorithm, SHA-256
+func (sign *Signature) buildStringToSign() {
+	sign.stringToSign = strings.Join([]string{
+		// Step 1: is to define the hashing Algorithm.
+		authHeaderPrefix,
+		// Step 2: Append the request date value which is specified with ISO8601 basic format
+		// in the x-auth-date header in the format YYYYMMDD'T'HHMMSS'Z'.
+		sign.requestDateTime,
+		//Step 3: Append the credential scope value that includes the date, the region,service,
+		// and a termination string ("sign_request") in lowercase characters.
+		sign.credentialString,
+		//Step 4: Append the hash of the canonical request created in Task 1
+		hex.EncodeToString(makeSha256([]byte(sign.canonicalString))),
```

Comment:
is it secure to store in String ? in JAVA it is recommended to store in char[], since due to GC not being done, the value of the String variable can be seen by a memory profiler